File: backend/app/schemas/survey.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class SurveyQuestionOption(BaseModel):
    text: str
    value: str | None = None


class SurveyQuestion(BaseModel):
    id: str
    text: str
    question_type: str = Field(
        ..., pattern=r"^(single_choice|multiple_choice|text|rating|slider)$"
    )
    options: list[SurveyQuestionOption] | None = None
    required: bool = True
    min_value: int | None = None
    max_value: int | None = None


class SurveyCreate(BaseModel):
    content_id: uuid.UUID
    title: str = Field(..., min_length=1, max_length=500)
    description: str | None = Field(default=None, max_length=2000)
    questions: list[SurveyQuestion] = Field(..., min_length=1)
    is_anonymous: bool = False


class SurveyUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=500)
    description: str | None = None
    questions: list[SurveyQuestion] | None = None
    is_anonymous: bool | None = None
    is_active: bool | None = None


class SurveyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    content_id: uuid.UUID
    title: str
    description: str | None = None
    questions: list[dict]
    response_count: int = 0
    is_anonymous: bool = False
    is_active: bool = True
    created_at: datetime
    updated_at: datetime


class SurveyListResponse(BaseModel):
    data: list[SurveyResponse]
    total: int
    page: int = 1
    pages: int = 1


class AnswerItem(BaseModel):
    question_id: str
    value: str | int | float | list[str] | None = None


class SurveySubmitRequest(BaseModel):
    employee_id: uuid.UUID | None = None
    responses: list[AnswerItem] = Field(..., min_length=1)


class SurveySubmitResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    survey_id: uuid.UUID
    employee_id: uuid.UUID | None = None
    responses: list[dict]
    submitted_at: datetime


class QuestionAggregation(BaseModel):
    question_id: str
    question_text: str
    question_type: str
    total_responses: int = 0
    distribution: dict[str, int] | None = None
    average: float | None = None
    text_responses: list[str] | None = None


class SurveyAggregationResponse(BaseModel):
    survey_id: uuid.UUID
    total_responses: int = 0
    questions: list[QuestionAggregation] = []

File: mobile/lib/models/survey.dart
/// Survey models for Session 73.

class SurveyData {
  final String id;
  final String title;
  final String? description;
  final List<SurveyQuestion> questions;
  final int responseCount;
  final bool isAnonymous;
  final bool isActive;

  const SurveyData({
    required this.id,
    required this.title,
    this.description,
    required this.questions,
    this.responseCount = 0,
    this.isAnonymous = false,
    this.isActive = true,
  });

  factory SurveyData.fromJson(Map<String, dynamic> json) {
    final questionsJson = json['questions'] as List? ?? [];
    return SurveyData(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      questions: questionsJson
          .map((q) => SurveyQuestion.fromJson(q as Map<String, dynamic>))
          .toList(),
      responseCount: json['response_count'] as int? ?? 0,
      isAnonymous: json['is_anonymous'] as bool? ?? false,
      isActive: json['is_active'] as bool? ?? true,
    );
  }

  int get requiredCount => questions.where((q) => q.required).length;
}

class SurveyQuestion {
  final String id;
  final String text;
  final String questionType;
  final List<SurveyOption> options;
  final bool required;
  final int? minValue;
  final int? maxValue;

  const SurveyQuestion({
    required this.id,
    required this.text,
    required this.questionType,
    this.options = const [],
    this.required = true,
    this.minValue,
    this.maxValue,
  });

  factory SurveyQuestion.fromJson(Map<String, dynamic> json) {
    final optionsJson = json['options'] as List? ?? [];
    return SurveyQuestion(
      id: json['id'] as String,
      text: json['text'] as String,
      questionType: json['question_type'] as String,
      options: optionsJson
          .map((o) => SurveyOption.fromJson(o as Map<String, dynamic>))
          .toList(),
      required: json['required'] as bool? ?? true,
      minValue: json['min_value'] as int?,
      maxValue: json['max_value'] as int?,
    );
  }
}

class SurveyOption {
  final String text;
  final String? value;

  const SurveyOption({required this.text, this.value});

  String get effectiveValue => value ?? text;

  factory SurveyOption.fromJson(Map<String, dynamic> json) {
    return SurveyOption(
      text: json['text'] as String,
      value: json['value'] as String?,
    );
  }
}

class SurveyAnswer {
  final String questionId;
  final dynamic value;

  const SurveyAnswer({required this.questionId, required this.value});

  Map<String, dynamic> toJson() => {
    'question_id': questionId,
    'value': value,
  };
}
